### PR TITLE
Bug 1839101: Add namespaced attribute to main navigation links

### PR DESCRIPTION
**Pull request by jerolimov**
**Fixes**:
https://issues.redhat.com/browse/ODC-3976
https://bugzilla.redhat.com/show_bug.cgi?id=1839101

Based on PR #5831 from @fmartingr - Big thanks for this work!!

**Analysis / Root cause**: 
Currently the navigation from the console works well when the target comes from a resource in the cluster, but some views that depends on namespaces that are not resources (like topology) use direct links that need to handle the namespace the user is on internally.

This causes a mild bug on the console when you have two tabs open on different namespaces, since these links use the local storage to determine the namespace the user is on and that would be the last one the user accessed.

**Solution Description**: 
Created a new navigation item that operates using a name and a namespace in the form of: <view>/ns/<namespace> which is the format of the navigation items replaced with this.

**Screen shots / Gifs for design review**: 
Before:
https://user-images.githubusercontent.com/139310/109855919-29ca7600-7c59-11eb-9c82-863ab0ab445b.mp4

After:
https://user-images.githubusercontent.com/139310/109855939-3058ed80-7c59-11eb-92b4-07dbcc47b53d.mp4

**Unit test coverage report**:
Tests not touched

**Test setup:**
* Open console in two tabs
* Navigate to project A in tab A
* Navigate to another project in tab B (switch project for example)
* Open navigation links in tab A which should open project A (opens latest used project from user settings before)

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Review by rottencandy — COMMENTED** [review]
Tested locally, looks good.
/lgtm

File: frontend/public/components/nav/items.tsx
Comment (christianvogt) [inline frontend/public/components/nav/items.tsx:185]: This should be optional. Defaults to `false`.
This would reduce the overall size of this PR since a lot of changes involve setting this value to `false`.
```suggestion
  namespaced?: boolean;
```

File: frontend/public/components/nav/admin-nav.tsx
Comment (christianvogt) [inline frontend/public/components/nav/admin-nav.tsx:81]: Default value for `namespaced` should be `false` and therefore this change won't be necessary.
```suggestion
```

File: frontend/packages/console-plugin-sdk/src/typings/nav-items.ts
Comment (christianvogt) [inline frontend/packages/console-plugin-sdk/src/typings/nav-items.ts:94]: Why are you removing this check? `isNavItem` is supposed to include all nav item extensions.

I do see an issue in `perspective-nav.tsx` though:

```
        if (isNavItem(item)) {
          return createLink(item, true);
        }
        if (isSeparatorNavItem(item)) {
          return <NavItemSeparator key={`separator-${index}`} />;
        }
```

Should be:
```
        if (isSeparatorNavItem(item)) {
          return <NavItemSeparator key={`separator-${index}`} />;
        }
        return createLink(item, true);
```

File: frontend/public/components/nav/admin-nav.tsx
Comment (jerolimov) [reply frontend/public/components/nav/admin-nav.tsx:81]: Ack. Solved

File: frontend/packages/console-plugin-sdk/src/typings/nav-items.ts
Comment (jerolimov) [reply frontend/packages/console-plugin-sdk/src/typings/nav-items.ts:94]: We can keep this as it was before. I just take this from the origin request but don't see a difference.

Comment (jerolimov) [reply frontend/packages/console-plugin-sdk/src/typings/nav-items.ts]: Applied your recommendation and rebased everything.

**Comment by jerolimov** [comment]
Removed all unnecessary `namespaced: false` and `namespaced={false}` values after making the new `namespaced` prop optional. Please take a new look.

File: frontend/public/components/nav/perspective-nav.tsx
Comment (rohitkrai03) [inline frontend/public/components/nav/perspective-nav.tsx:119]: @christianvogt's suggestion was to remove the check for `isNavItem` and directly return `createLink` in the end since `isNavItem` would be true for all nav extensions.

```suggestion
        if (isNavItem(item)) {
          return createLink(item, true);
        }
```

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

**Comment by jerolimov** [comment]
Rebased and just fixed the last comment of Rohit https://github.com/openshift/console/pull/8249#pullrequestreview-616537370

**Review by rohitkrai03 — APPROVED** [review]
/approve

**Review by rottencandy — APPROVED** [review]
/lgtm

**Comment by glekner** [comment]
/hold

this conflicts with https://github.com/openshift/console/pull/8438
lets not merge code thats going to be deleted anyway.

**Comment by rohitkrai03** [comment]
I agree. Since we are already migrating to dynamic plugin extensions, all the static extensions and their usage are going to be deleted anyway. Let's work to get these changes in the dynamic extensions PR.

**Comment by andrewballantyne** [comment]
This PR seems to have stalled for 22 days heh. Are we looking to keep this or merge it in with another PR?

cc @rohitkrai03 @jerolimov

**Comment by jerolimov** [comment]
@andrewballantyne We wait for #8438. I'm super happy to close this here, but I don't want do this without the other one is merged.

@glekner will you have time to work on, or at least rebase #8438 soon or should we merge this one here and hope that your PR do not lose this feature again?

**Comment by andrewballantyne** [comment]
@glekner @jerolimov We definitely should resolve this one way or another. If we are not pushing for the dependant PR to get merged, we should either close this (and make it part of the dependant PR down the road) or merge this without it. ODC shouldn't stall a PR for a month because of a dependant. We need to get better at managing our tickets and PRs, and this is going to drag our metrics down.

I'll leave it up to you guys, if you want to try to merge this in the intended order, please make concerted efforts to do so. If not, we need to look at what to do about this PR for our own team. Bugs impact the quality of where our team is at through the new SLO metrics, and this PR is weighing down the team because of its existence. So something needs to happen here.

cc @nimishamukherjee @christianvogt @debsmita1

**Comment by glekner** [comment]
/unhold

sorry for the late response as I was busy with bugs. let's merge this one, ill deal with the rebases.

**Comment by jerolimov** [comment]
All fine, thanks Gilad!

**Pull request merged**
